code/modules/atmospherics/machinery/pipes/layermanifold.dm

```diff
@@ -0,0 +1,129 @@
+	for(var/obj/machinery/atmospherics/A in all_connected)
+		returnlist += all_connected[A]
+	return returnlist
+
+/obj/machinery/atmospherics/pipe/layer_manifold/Destroy()
+	nullifyAllNodes()
+	..()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/update_icon()	//HEAVILY WIP FOR UPDATE ICONS!!
+	layer = (initial(layer) + (PIPING_LAYER_MAX * PIPING_LAYER_LCHANGE))	//This is above everything else.
+	var/invis = invisibility ? "-f" : ""
+	icon_state = "[initial(icon_state)][invis]"
+	underlays.Cut()
+	for(var/obj/machinery/atmospherics/A in front_nodes)
+		add_attached_image(A)
+	for(var/obj/machinery/atmospherics/A in back_nodes)
+		add_attached_image(A)
+
+/obj/machinery/atmospherics/pipe/layer_manifold/proc/add_attached_image(obj/machinery/atmospherics/A)
+	var/invis = A.invisibility ? "-f" : ""
+	if(istype(A, /obj/machinery/atmospherics/pipe/layer_manifold))
+		for(var/i = PIPING_LAYER_MIN, i <= PIPING_LAYER_MAX, i++)
+			var/image/I = getpipeimage('icons/obj/atmospherics/pipes/manifold.dmi', "manifold_full_long[invis]", get_dir(src, A))
+			I.pixel_x = (i - PIPING_LAYER_DEFAULT) * PIPING_LAYER_P_X
+			I.pixel_y = (i - PIPING_LAYER_DEFAULT) * PIPING_LAYER_P_Y
+			underlays += I
+	else
+		var/image/I = getpipeimage('icons/obj/atmospherics/pipes/manifold.dmi', "manifold_full_long[invis]", get_dir(src, A))
+		I.pixel_x = A.pixel_x
+		I.pixel_y = A.pixel_y
+		underlays += I
+
+/obj/machinery/atmospherics/pipe/layer_manifold/SetInitDirections()
+	switch(dir)
+		if(NORTH || SOUTH)
+			initialize_directions = NORTH|SOUTH
+		if(EAST || WEST)
+			initialize_directions = EAST|WEST
+
+/obj/machinery/atmospherics/pipe/layer_manifold/isConnectable(obj/machinery/atmospherics/target, given_layer)
+	if(!given_layer)
+		return TRUE
+	. = ..()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/proc/findAllConnections()
+	front_nodes = list()
+	back_nodes = list()
+	var/list/new_nodes = list()
+	for(var/iter = PIPING_LAYER_MIN; iter <= PIPING_LAYER_MAX; iter += 1)
+		var/obj/machinery/atmospherics/foundfront = findConnecting(dir, iter)
+		var/obj/machinery/atmospherics/foundback = findConnecting(turn(dir, 180), iter)
+		front_nodes += foundfront
+		back_nodes += foundback
+		if(foundfront && !QDELETED(foundfront))
+			new_nodes += foundfront
+		if(foundback && !QDELETED(foundback))
+			new_nodes += foundback
+	update_icon()
+	return new_nodes
+
+/obj/machinery/atmospherics/pipe/layer_manifold/atmosinit()
+	findAllConnections()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/setPipingLayer(new_layer = PIPING_LAYER_DEFAULT)
+	piping_layer = PIPING_LAYER_DEFAULT
+
+/obj/machinery/atmospherics/pipe/layer_manifold/pipeline_expansion()
+	return get_all_connected_nodes()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/disconnect(obj/machinery/atmospherics/reference)
+	if(istype(reference, /obj/machinery/atmospherics/pipe))
+		var/obj/machinery/atmospherics/pipe/P = reference
+		P.destroy_network()
+	while(reference in get_all_connected_nodes())
```

Comment:
uhm why not get this as a list once and then loop over it? this is rebuilding a list over and over which is pretty ugly, is there no better way to do this?

why not book keep an allnodes list (updated whenever somethign is added/remove to nodes/frontnodes or backnodes) and then use that